IPython/core/events.py

```diff
@@ -90,29 +125,53 @@ def _define_event(callback_proto):
 # ------------------------------------------------------------------------------
 
 @_define_event
-def pre_execute():
+def pre_execute(result):
```

Comment:
It seems strange for the *pre* events to carry a result object. Also, the code in `interactiveshell` that triggers these events does not give them the result.

Reply:
Opps, looks like I forgot to provide the result as an argument. Anyway, if you don't want the *pre* events that carry as result object, I can clean that up again. I just thought it would be useful, so user for example can "memorize" a result in a *pre* event and then match them in the corresponding *post* event (stupid example: pre event allocates memory for a computation, post event frees that corresponding memory allocation again if that makes sense).
Remove or keep the result argument?

Reply:
Oh, just using it as a handle to match up events? I guess so, but I find it hard to think of a real case where you'd want that. @minrk?

Reply:
I would expect the `pre` events would take the requested code to be executed (basically the ExecuteRequest from the message perspective), i.e. a dict of `run_cell`'s kwargs. I've encountered several places where people want pre-execute notifiers of executions about to start.

For matching things up, since execution is ordered, there should always be one post_execute after each pre_execute, never two pre_executes in a row, so I'm not sure there is a use case for the pre_execute getting the incomplete result object. I, too, found it a little surprising to have a `result` object prior to the execution occurring.

Reply:
Thanks for the comments. I've introduced `ExecuteRequest` which does what @minrk suggested. Additionally, I've added a reference to the corresponding `ExecuteRequest` to `ExecuteResult`s which allows to retrieve for example the cell code in *post* events as well (and also to match up events if this is actually a use case).